pkg/actions/cluster/unowned.go

```diff
@@ -21,3 +36,198 @@ func (c *UnownedCluster) Upgrade(dryRun bool) error {
+	}
+
+	// we have to wait for nodegroups to delete before deleting the cluster
+	// so the `wait` value is ignored here
+	if err := c.deleteAndWaitForNodegroupsDeletion(clusterName, waitTimeout); err != nil {
+		return err
+	}
+
+	if err := c.deleteIAMAndOIDC(wait, kubernetes.NewCachedClientSet(clientSet)); err != nil {
+		return err
+	}
+
+	return c.deleteCluster(clusterName, waitTimeout, wait)
+}
+
+func (c *UnownedCluster) checkClusterExists(clusterName string) error {
+	out, err := c.ctl.Provider.EKS().ListClusters(&awseks.ListClustersInput{})
+	if err != nil {
+		return err
+	}
+	for _, cluster := range out.Clusters {
+		if *cluster == clusterName {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("cluster %q not found", clusterName)
+}
+
+func (c *UnownedCluster) deleteIAMAndOIDC(wait bool, clientSetGetter kubernetes.ClientSetGetter) error {
+	clusterOperable, _ := c.ctl.CanOperate(c.cfg)
+
+	var oidc *iamoidc.OpenIDConnectManager
+	var err error
+	stackManager := c.ctl.NewStackManager(c.cfg)
+
+	oidcSupported := true
+	if clusterOperable {
+		oidc, err = c.ctl.NewOpenIDConnectManager(c.cfg)
+		if err != nil {
+			if _, ok := err.(*eks.UnsupportedOIDCError); !ok {
+				return err
+			}
+			oidcSupported = false
+		}
+	}
+
+	deleteOIDCProvider := clusterOperable && oidcSupported
+
+	tasks := &manager.TaskTree{Parallel: false}
+
+	if deleteOIDCProvider {
+		serviceAccountAndOIDCTasks, err := stackManager.NewTasksToDeleteOIDCProviderWithIAMServiceAccounts(oidc, clientSetGetter)
+		if err != nil {
+			return err
+		}
+
+		if serviceAccountAndOIDCTasks.Len() > 0 {
+			serviceAccountAndOIDCTasks.IsSubTask = true
+			tasks.Append(serviceAccountAndOIDCTasks)
+		}
+	}
+
+	deleteAddonIAMtasks, err := stackManager.NewTaskToDeleteAddonIAM(wait)
+	if err != nil {
+		return err
+	}
+
+	if deleteAddonIAMtasks.Len() > 0 {
+		deleteAddonIAMtasks.IsSubTask = true
+		tasks.Append(deleteAddonIAMtasks)
+	}
+
+	if tasks.Len() == 0 {
+		logger.Warning("no IAM and OIDC resources were found for %q", c.cfg.Metadata.Name)
+		return nil
+	}
+
+	logger.Info(tasks.Describe())
+	if errs := tasks.DoAllSync(); len(errs) > 0 {
+		return handleErrors(errs, "deleting cluster IAM and OIDC")
+	}
+
+	logger.Info("all IAM and OIDC resources were deleted")
+	return nil
+}
+
+func (c *UnownedCluster) deleteCluster(clusterName string, waitTimeout time.Duration, wait bool) error {
+	out, err := c.ctl.Provider.EKS().DeleteCluster(&awseks.DeleteClusterInput{
+		Name: &clusterName,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	logger.Info("initiated deletion of cluster %q", clusterName)
+	if out != nil {
+		logger.Debug("delete cluster response: %s", out.String())
+	}
+
+	if !wait {
+		logger.Info("to see the status of the deletion run `eksctl get cluster --name %s --region %s`", clusterName, c.cfg.Metadata.Region)
+		return nil
+	}
+
+	condition := func() (bool, error) {
+		clusterDeleted := true
+		clusters, err := c.ctl.Provider.EKS().ListClusters(&awseks.ListClustersInput{})
```

Comment:
can you just `GetCluster` and success on 404? Is pagination handled correctly here?

Reply:
👍 good spot